### PR TITLE
RPYrT Support. Format checked.

### PR DESCRIPTION
**Pull request by Pedro-Roque**
**Test data / coverage**
Tested in SITL and flight arena. Here goes a link to the folder with the logs:
https://drive.google.com/open?id=1Ja1JjI3Ggmw383kDTZMdCB0cr3Hd5gfp

**Setpoints:** All data exchange happens through a SiK Radio (433Mhz), which is way less than ideal for the rates at which we are controlling - I am publishing at 100Hz, but I am pretty sure it can't handle such rates. Iris vehicle with PX4-FMU_v2. These use LPE estimator (these tests were made November 2018). It looks good, except for some points in which the telemetry radio looses connection (makes the UAV drop), but it recovers from it pretty fine. Both position and attitude changes are tested. It seems that the LPE sometimes drops to 0.2m in Z - is there any param I am missing to set? I am running this on an arena with MoCap. 

**TestWithDisturbances:** all data exchanged happends through a Mav-esp with custom firmware. The control frequency is 50Hz, while odometry is fed at 100Hz. Vehicle running Pixracer with the same changes present in this PR. The logs resamble 3 tests on a real UAV performed on a flight arena with Qualisys Mocap. RPYrT setpoints are externally given to the UAV via offboard mode by an in-house developed PID offboard controller, and external unmodeled disturbances are applied to the vehicles. The chain of control is: position setpoint --> PID controller (Offboard) --> RPYrT setpoint --> Attitude controller (onboard).

**Describe problem solved by the proposed pull request**
Essentially fixes the ignore_bodyrate for all axes, and adds functionality for single-axis only. At the moment, it targets just yawrate control mode.

**Describe your preferred solution**
Added offboard control mode yawrate, which ignores x and y rates, and uses the z rate as the setpoint for the rate controller. Roll and Pitch setpoints are still used for the attitude control to set the desired UAV attitude. At the moment, the yaw angle is kept to the same one that the UAV has (suggestions for this are welcome!)

**Describe possible alternatives**
-

**Additional context**
Flight tests show that RPYrT setpoints work as expected, as the UAV successfully converges to the right setpoints. Tested successfully in SITL as well.

### CODE REVIEW

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 1]
```diff
@@ -918,8 +918,10 @@ MavlinkReceiver::handle_message_set_position_target_local_ned(mavlink_message_t
 		/* yaw ignore flag mapps to ignore_attitude */
 		offboard_control_mode.ignore_attitude = (bool)(set_position_target_local_ned.type_mask & 0x400);
 		/* yawrate ignore flag mapps to ignore_bodyrate */
-		offboard_control_mode.ignore_bodyrate = (bool)(set_position_target_local_ned.type_mask & 0x800);
-
+		// offboard_control_mode.ignore_bodyrate = (bool)(set_position_target_local_ned.type_mask & 0x800);
+		offboard_control_mode.ignore_bodyrate_x = (bool)(set_position_target_local_ned.type_mask & 0x800);
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp]: Can we define these masks somewhere?

Comment (dagar) [reply src/modules/mavlink/mavlink_receiver.cpp]: It should really be part of the mavlink spec, but at least for our own sanity we could define them in PX4.

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 2]
```diff
@@ -895,6 +895,15 @@ MulticopterAttitudeControl::run()
 							!_v_control_mode.flag_control_position_enabled) {
 						generate_attitude_setpoint(attitude_dt, reset_yaw_sp);
 						attitude_setpoint_generated = true;
+					} 
+
+					if (_v_control_mode.flag_control_yawrate_override_enabled)
+					{
```
Comment (dagar) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: formatting

**Comment by dagar** [comment]
I'm not sure about this one. If I understand the change correctly aren't we going to have multiple rate setpoints publishing simultaneously?

**Comment by Pedro-Roque** [comment]
@dagar , I don't thnink so, as the yaw rate on the Attitude controller is essentially changed to the yaw rate coming from offboard - gets overriden. In simulation it checks out, primary tests also checked out (but some oscillation after about 15s around the setpoint, but this I have an idea on what might cause it.

I plan to do some more flights this week and the next. If all checks out, I will pust logs and videos here.

**Comment by Pedro-Roque** [comment]
@dagar Finally made it to get some time for flight tests. Updated the first PR message with a folder to the logs.

TL;DR: rpyt looks fine!

**Comment by Pedro-Roque** [comment]
@dagar  I have more flight tests + working version merged with latest master v1.9 . Let me know if I should push it to this branch to be merged

**Comment by MaEtUgR** [comment]
Discussed in the dev call. @Pedro-Roque rebases on 1.9 master and we take a look and do tests such that we can merge it.

**Comment by MaEtUgR** [comment]
@Pedro-Roque Thanks for rebasing 😉

**Comment by Pedro-Roque** [comment]
> @Pedro-Roque Thanks for rebasing

Thank you for helping on it! Updated log folder and PR description. Let me know if there is anything else I can help with

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 3]
```diff
@@ -833,8 +833,10 @@ MavlinkReceiver::handle_message_set_position_target_local_ned(mavlink_message_t
 		/* yaw ignore flag mapps to ignore_attitude */
 		offboard_control_mode.ignore_attitude = (bool)(set_position_target_local_ned.type_mask & 0x400);
 		/* yawrate ignore flag mapps to ignore_bodyrate */
-		offboard_control_mode.ignore_bodyrate = (bool)(set_position_target_local_ned.type_mask & 0x800);
-
+		// offboard_control_mode.ignore_bodyrate = (bool)(set_position_target_local_ned.type_mask & 0x800);
```
Comment (mhkabir) [inline src/modules/mavlink/mavlink_receiver.cpp]: Remove commented code.

File: msg/offboard_control_mode.msg [hunk 4]
```diff
@@ -5,6 +5,9 @@ uint64 timestamp		# time since system start (microseconds)
 bool ignore_thrust
 bool ignore_attitude
 bool ignore_bodyrate
```
Comment (mhkabir) [inline msg/offboard_control_mode.msg]: Remove? Should be unused now.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 5]
```diff
@@ -1383,13 +1401,22 @@ MavlinkReceiver::handle_message_set_attitude_target(mavlink_message_t *msg)
 
 				/* Publish attitude rate setpoint if bodyrate and thrust ignore bits are not set */
 				///XXX add support for ignoring individual axes
```
Comment (mhkabir) [inline src/modules/mavlink/mavlink_receiver.cpp]: Comment now unnecessary

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 6]
```diff
@@ -834,9 +834,17 @@ MulticopterAttitudeControl::run()
 							!_v_control_mode.flag_control_position_enabled) {
 						generate_attitude_setpoint(attitude_dt, reset_yaw_sp);
 						attitude_setpoint_generated = true;
-					}
+					}
```
Comment (mhkabir) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: Whitespace

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 5 above)
Comment (Pedro-Roque) [reply src/modules/mavlink/mavlink_receiver.cpp]: Not sure if unnecessary. I implemented Yawrate control, not for the other axis though. Still WIP

File: msg/offboard_control_mode.msg (hunk 4 above)
Comment (Pedro-Roque) [reply msg/offboard_control_mode.msg]: you mean ignore_bodyrate? Yes can check on that and remove it

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 6 above)
Comment (Pedro-Roque) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: Fixed.

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 3 above)
Comment (Pedro-Roque) [reply src/modules/mavlink/mavlink_receiver.cpp]: Fixed. also removed l.835

File: msg/offboard_control_mode.msg (hunk 4 above)
Comment (Pedro-Roque) [reply msg/offboard_control_mode.msg]: Fixed. Removed it and build checks out.

**Comment by Pedro-Roque** [comment]
@MaEtUgR wierd VTOL SITL error. Re-run, or is there something wrong with it?

**Comment by MaEtUgR** [comment]
I created a copy to untangle the changes from existing commits on master: #12061

**Comment by julianoes** [comment]
@MaEtUgR ok what am I reviewing?

**Comment by Pedro-Roque** [comment]
> @MaEtUgR ok which pull request should I now review?

I think the now that @MaEtUgR  created, @julianoes .
